src/System.Reflection.Metadata.ILDasm/src/Decoder/ILDecoder.cs

```diff
@@ -0,0 +1,582 @@
+                        break;
+                    case OperandType.InlineSwitch:
+                        instruction = CreateSwitchInstruction(ref ilReader, expectedSize, ilOffset, opCode);
+                        break;
+                    case OperandType.ShortInlineBrTarget:
+                        instruction = new ILInt16BranchInstruction(opCode, ilReader.ReadSByte(), ilOffset, expectedSize + 1);
+                        break;
+                    case OperandType.InlineBrTarget:
+                        instruction = new ILBranchInstruction(opCode, ilReader.ReadInt32(), ilOffset, expectedSize + 4);
+                        break;
+                    case OperandType.ShortInlineI:
+                        instruction = new ILByteInstruction(opCode, ilReader.ReadByte(), -1, expectedSize + 1);
+                        break;
+                    case OperandType.ShortInlineR:
+                        instruction = new ILSingleInstruction(opCode, ilReader.ReadSingle(), -1, expectedSize + 4);
+                        break;
+                    case OperandType.InlineNone:
+                        instruction = new ILInstructionWithNoValue(opCode, expectedSize);
+                        break;
+                    case OperandType.ShortInlineVar:
+                        byte token = ilReader.ReadByte();
+                        instruction = new ILInt16VariableInstruction(opCode, GetVariableName(opCode, token, _methodDefinition), token, expectedSize + 1);
+                        break;
+                    case OperandType.InlineVar:
+                        shortOperand = ilReader.ReadUInt16();
+                        instruction = new ILVariableInstruction(opCode, GetVariableName(opCode, shortOperand, _methodDefinition), shortOperand, expectedSize + 2);
+                        break;
+                    case OperandType.InlineSig:
+                        intOperand = ilReader.ReadInt32();
+                        instruction = new ILStringInstruction(opCode, GetSignature(mdReader, intOperand, provider), intOperand, expectedSize + 4);
+                        break;
+                    default:
+                        break;
+                }
+                ilOffset += instruction.Size;
+                yield return instruction;
+            }
+        }
+
+        internal static ILLocal[] DecodeLocalSignature(MethodBodyBlock methodBody, MetadataReader mdReader, ILTypeProvider provider)
+        {
+            if (methodBody.LocalSignature.IsNil)
+            {
+                return new ILLocal[0];
+            }
+            var localTypes = SignatureDecoder.DecodeLocalSignature(methodBody.LocalSignature, provider);
+            ILLocal[] locals = new ILLocal[localTypes.Count()];
+            for (int i = 0; i < localTypes.Length; i++)
+            {
+                string name = "V_" + i;
+                locals[i] = new ILLocal(name, localTypes[i].ToString());
```

Comment:
Looks like we're throwing away some information by converting the type to a string.  We don't want ILLocal to store the string name and the strongly-typed type?  Or we can't for some reason?